### DISCUSSION TITLE
Typesafety in sendTo action and fromCallback actor

### DISCUSSION THREAD

**Discussion by Veronika11111** [statelyai/xstate#4995] [Q&A]
I have simple machine which opens and closes the gate (it is just piece of bigger project). This machine invokes the GateActor, which is responsible for communication with HW gate. This actor is fromCallback actor. I have 2 problems with typescript here:

FIRST PROBLEM
- when calling sendBack function from the GateActor, the event type is not type-safe. I understand that it is because the parent of the actor can be different and in other chats you advice to pass parent as input for GateActor (check [here](https://github.com/statelyai/xstate/issues/4666)). Can you please provide me an example of how to do it in my case?

SECOND PROBLEM
- when I call sendTo function in the action of the parent machine, typescript can not check if the event type which I am sending to GateActor really exists in the GateActor. I can not find any solution for this in discussions. 

The type safety is important for me, because it is easy to create a typo in event name and it is difficult to debug it then.

Thank you in advance.

```ts
export const gateMachine = setup({
  types: {
    context: {} as { gate: IGate },
    events: {} as
       | { type: 'user.openGate' }
       | { type: 'gate.move_done' }
       | { type: 'user.close_gate' };,
    input: {} as { gate: IGate },
  },

  actors: {
    GateActor: fromCallback<EventObject, { gate: IGate }>(({ sendBack, receive, input }) => {
      input.gate.addEventListener(event => {
        if (event === GateEventType.MOVE_DONE) {
// FIRST PROBLEM
          sendBack({ type: 'gate.move_done' });
        }
      });
    
      receive(event => {
        assertEvent(event, ['open', 'close']);
        switch (event.type) {
          case 'open':
            input.gate.openGate();
            break;
          case 'close':
            input.gate.closeGate();
            break;
        }
      });
    
      return () => {
        // cleanup
        input.gate.removeEventListener();
      };
    }),
  },
  actions: {
// SECOND PROBLEM
    openGate: sendTo('gateActor', { type: 'open' }),
    closeGate: sendTo('gateActor', {
      type: 'close',
    }),
  },
}).createMachine({
  context: ({ input }) => ({
    gate: input?.gate,
  }),
  id: 'Only Gate simulator',
  initial: 'Ready',
  invoke: {
    id: 'gateActor',
    input: ({ context }) => ({ gate: context.gate }),
    src: 'GateActor',
  },
  states: {
    Ready: {
      on: {
        'user.openGate': {
          target: 'OpeningGate',
        },
      },
    },
    OpeningGate: {
      on: {
        'gate.move_done': {
          target: 'AwaitingPackage',
        },
      },
      entry: {
        type: 'openGate',
      },
    },
    AwaitingPackage: {
      on: {
        'user.close_gate': {
          target: 'Ready',
        },
      },
      exit: {
        type: 'closeGate',
      },
    },
  },
});
```

**Comment by eug-vs** [comment]
I worked around this by a dirty typecast, basically making the `sendTo` believe that we pass actorRef into it, not just an ID string.
```typescript
  actions: sendTo(
    'myActorId' as unknown as ActorRefFromLogic<typeof actorLogic>,
    event, // This is now type-safe
  ),
```
In this case actor logic can be extracted into a variable (outside of machine) `const actorLogic = fromCallback(...)` and then used to improve inference
